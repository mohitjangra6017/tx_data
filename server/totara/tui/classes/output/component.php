<?php
/*
 * This file is part of Totara Learn
 *
 * Copyright (C) 2019 onwards Totara Learning Solutions LTD
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * @author Simon Chester <simon.chester@totaralearning.com>
 * @package totara_core
 */

namespace totara_tui\output;

/**
 * Renderable TUI component
 */
final class component implements \renderable {
    /**
     * @var string Component name
     */
    private $name;

    /**
     * @var array|null Component props
     */
    private $props = [];

    /**
     * @var bool
     */
    private static $registered = [];

    /**
     * Create a new instance of component
     *
     * @param string $name Module ID of the component to render.
     *     e.g. 'totara_core/components/Example`.
     * @param array|null $props Props to pass to the component.
     *     Props are not processed or escaped in any way, they are passed as-is
     *     to the component. Must be encodable as JSON.
     */
    public function __construct(string $name, array $props = null) {
        $this->name = $name;
        $this->props = $props;
    }

    /**
     * Get component name
     *
     * @return string
     */
    public function get_name(): string {
        return $this->name;
    }

    /**
     * Get component props
     *
     * @return array Mapping of prop keys to values
     */
    public function get_props(): ?array {
        return $this->props;
    }

    /**
     * @return bool
     */
    public function has_props(): bool {
        return ($this->props !== null);
    }

    /**
     * @throws \coding_exception if there are problems json encoding data for the component.
     * @return string
     */
    public function get_props_encoded(): string {
        if (!$this->has_props()) {
            // If you get here and want to remove this then you must work out and confirm what empty props look like for Vue.
            // It wasn't required during development and it's a bad idea to support something you don't need!
            throw new \coding_exception('Encoded props requested, but there are no props.');
        }
        $data = json_encode($this->props);
        if ($data === false && json_last_error() !== JSON_ERROR_NONE) {
            throw new \coding_exception('Invalid component data encountered while attempting to encode a component of type ' . static::class, json_last_error_msg());
        }
        return $data;
    }

    public function register(\moodle_page $page): \moodle_page {
        return self::register_component($this->get_name(), $page);
    }

    public static function register_component(string $name, \moodle_page $page): \moodle_page {
        $pos = strpos($name, '/');
        if ($pos !== false) {
            $name = substr($name, 0, $pos);
        }
        if (!isset(self::$registered[$name])) {
            if ($page->state >= \moodle_page::STATE_IN_BODY) {
                throw new \coding_exception('Unable to register component as the header has already been printed.');
            }
            self::get_framework($page)->require_vue($name);
            self::$registered[$name] = true;
        }
        return $page;
    }

    /**
     * @param \moodle_page $page
     * @throws \coding_exception
     * @return framework
     */
    private static function get_framework(\moodle_page $page): framework {
        /** @var framework $framework */
        $framework = $page->requires->framework(framework::class);
        return $framework;
    }
}
