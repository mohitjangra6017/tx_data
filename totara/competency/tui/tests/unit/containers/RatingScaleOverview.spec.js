/*
 * This file is part of Totara Learn
 *
 * Copyright (C) 2020 onwards Totara Learning Solutions LTD
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * @author Mark Metcalfe <mark.metcalfe@totaralearning.com>
 * @package totara_competency
 */

import { shallowMount } from '@vue/test-utils';
import component from 'totara_competency/components/RatingScaleOverview';
import { mocks } from './mocks';

const props = {
  reverseValues: false,
  scale: {
    values: [
      {
        id: '6',
        name: 'Extremely competent',
        proficient: true,
        description: '<b>No doubt this fella is competent</b>',
      },
      {
        id: '7',
        name: 'Competent',
        proficient: true,
        description: '<i>There is some merit co call it competent</i>',
      },
      {
        id: '8',
        name: 'Competent on Tuesdays',
        proficient: true,
        description: 'Competent, but only on Tuesdays, do not ask why.',
      },
      {
        id: '9',
        name: 'Not competent on Tuesdays',
        proficient: true,
        description: 'Not competent, but only on Tuesdays, do not ask why.',
      },
      {
        id: '10',
        name: 'Below average',
        proficient: false,
        description:
          'We can not call it competent for just yet, maybe wait for Tuesday.',
      },
      {
        id: '11',
        name: 'Not competent',
        proficient: false,
        description: 'Why does this value even exist?',
      },
    ],
  },
};

describe('components/RatingScaleOverview.vue', () => {
  it('Checks snapshot - with descriptions', () => {
    let wrapper = shallowMount(component, {
      mocks: mocks,
      propsData: Object.assign(props, {
        showDescriptions: true,
      }),
    });
    expect(wrapper.element).toMatchSnapshot('withDescriptions');
  });
  it('Checks snapshot - without descriptions', () => {
    let wrapper = shallowMount(component, {
      mocks: mocks,
      propsData: Object.assign(props, {
        showDescriptions: false,
      }),
    });
    expect(wrapper.element).toMatchSnapshot('withoutDescriptions');
  });
});
