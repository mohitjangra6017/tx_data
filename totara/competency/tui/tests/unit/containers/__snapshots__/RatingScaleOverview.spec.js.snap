// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/RatingScaleOverview.vue Checks snapshot - with descriptions: withDescriptions 1`] = `
<div
  class="tui-competencyRatingScaleOverview"
>
  <h3
    class="tui-competencyRatingScaleOverview__title"
    id="id-rating-scale-title"
  >
    
    $str('rating_scale', 'totara_competency')
  
  </h3>
   
  <ol
    aria-labelledby="id-rating-scale-title"
    class="tui-competencyRatingScaleOverview__list"
  >
    <li
      class="tui-competencyRatingScaleOverview__list-item"
    >
      <div
        class="tui-competencyRatingScaleOverview__proficientStatus"
      >
        <!---->
      </div>
       
      <div
        class="tui-competencyRatingScaleOverview__scaleValue"
      >
        <h4
          class="tui-competencyRatingScaleOverview__scaleValue-name"
        >
          
          Extremely competent
        
        </h4>
         
        <p
          class="tui-competencyRatingScaleOverview__scaleValue-description"
        >
          <b>
            No doubt this fella is competent
          </b>
        </p>
         
        <span
          class="sr-only"
        >
          
          $str('is_proficient_value', 'totara_competency')
        
        </span>
      </div>
    </li>
    <li
      class="tui-competencyRatingScaleOverview__list-item"
    >
      <div
        class="tui-competencyRatingScaleOverview__proficientStatus"
      >
        <!---->
      </div>
       
      <div
        class="tui-competencyRatingScaleOverview__scaleValue"
      >
        <h4
          class="tui-competencyRatingScaleOverview__scaleValue-name"
        >
          
          Competent
        
        </h4>
         
        <p
          class="tui-competencyRatingScaleOverview__scaleValue-description"
        >
          <i>
            There is some merit co call it competent
          </i>
        </p>
         
        <span
          class="sr-only"
        >
          
          $str('is_proficient_value', 'totara_competency')
        
        </span>
      </div>
    </li>
    <li
      class="tui-competencyRatingScaleOverview__list-item"
    >
      <div
        class="tui-competencyRatingScaleOverview__proficientStatus"
      >
        <!---->
      </div>
       
      <div
        class="tui-competencyRatingScaleOverview__scaleValue"
      >
        <h4
          class="tui-competencyRatingScaleOverview__scaleValue-name"
        >
          
          Competent on Tuesdays
        
        </h4>
         
        <p
          class="tui-competencyRatingScaleOverview__scaleValue-description"
        >
          Competent, but only on Tuesdays, do not ask why.
        </p>
         
        <span
          class="sr-only"
        >
          
          $str('is_proficient_value', 'totara_competency')
        
        </span>
      </div>
    </li>
    <li
      class="tui-competencyRatingScaleOverview__list-item"
    >
      <div
        class="tui-competencyRatingScaleOverview__proficientStatus"
      >
        <checksuccess-stub
          size="300"
        />
      </div>
       
      <div
        class="tui-competencyRatingScaleOverview__scaleValue"
      >
        <h4
          class="tui-competencyRatingScaleOverview__scaleValue-name"
        >
          
          Not competent on Tuesdays
        
        </h4>
         
        <p
          class="tui-competencyRatingScaleOverview__scaleValue-description"
        >
          Not competent, but only on Tuesdays, do not ask why.
        </p>
         
        <span
          class="sr-only"
        >
          
          $str('is_proficient_value', 'totara_competency')
        
        </span>
      </div>
    </li>
    <li
      class="tui-competencyRatingScaleOverview__list-item"
    >
      <div
        class="tui-competencyRatingScaleOverview__proficientStatus"
      >
        <!---->
      </div>
       
      <div
        class="tui-competencyRatingScaleOverview__scaleValue"
      >
        <h4
          class="tui-competencyRatingScaleOverview__scaleValue-name"
        >
          
          Below average
        
        </h4>
         
        <p
          class="tui-competencyRatingScaleOverview__scaleValue-description"
        >
          We can not call it competent for just yet, maybe wait for Tuesday.
        </p>
         
        <span
          class="sr-only"
        >
          
          $str('is_not_proficient_value', 'totara_competency')
        
        </span>
      </div>
    </li>
    <li
      class="tui-competencyRatingScaleOverview__list-item"
    >
      <div
        class="tui-competencyRatingScaleOverview__proficientStatus"
      >
        <!---->
      </div>
       
      <div
        class="tui-competencyRatingScaleOverview__scaleValue"
      >
        <h4
          class="tui-competencyRatingScaleOverview__scaleValue-name"
        >
          
          Not competent
        
        </h4>
         
        <p
          class="tui-competencyRatingScaleOverview__scaleValue-description"
        >
          Why does this value even exist?
        </p>
         
        <span
          class="sr-only"
        >
          
          $str('is_not_proficient_value', 'totara_competency')
        
        </span>
      </div>
    </li>
  </ol>
</div>
`;

exports[`components/RatingScaleOverview.vue Checks snapshot - without descriptions: withoutDescriptions 1`] = `
<div
  class="tui-competencyRatingScaleOverview"
>
  <h3
    class="tui-competencyRatingScaleOverview__title"
    id="id-rating-scale-title"
  >
    
    $str('rating_scale', 'totara_competency')
  
  </h3>
   
  <ol
    aria-labelledby="id-rating-scale-title"
    class="tui-competencyRatingScaleOverview__list"
  >
    <li
      class="tui-competencyRatingScaleOverview__list-item"
    >
      <div
        class="tui-competencyRatingScaleOverview__proficientStatus"
      >
        <!---->
      </div>
       
      <div
        class="tui-competencyRatingScaleOverview__scaleValue"
      >
        <h4
          class="tui-competencyRatingScaleOverview__scaleValue-name"
        >
          
          Extremely competent
        
        </h4>
         
        <!---->
         
        <span
          class="sr-only"
        >
          
          $str('is_proficient_value', 'totara_competency')
        
        </span>
      </div>
    </li>
    <li
      class="tui-competencyRatingScaleOverview__list-item"
    >
      <div
        class="tui-competencyRatingScaleOverview__proficientStatus"
      >
        <!---->
      </div>
       
      <div
        class="tui-competencyRatingScaleOverview__scaleValue"
      >
        <h4
          class="tui-competencyRatingScaleOverview__scaleValue-name"
        >
          
          Competent
        
        </h4>
         
        <!---->
         
        <span
          class="sr-only"
        >
          
          $str('is_proficient_value', 'totara_competency')
        
        </span>
      </div>
    </li>
    <li
      class="tui-competencyRatingScaleOverview__list-item"
    >
      <div
        class="tui-competencyRatingScaleOverview__proficientStatus"
      >
        <!---->
      </div>
       
      <div
        class="tui-competencyRatingScaleOverview__scaleValue"
      >
        <h4
          class="tui-competencyRatingScaleOverview__scaleValue-name"
        >
          
          Competent on Tuesdays
        
        </h4>
         
        <!---->
         
        <span
          class="sr-only"
        >
          
          $str('is_proficient_value', 'totara_competency')
        
        </span>
      </div>
    </li>
    <li
      class="tui-competencyRatingScaleOverview__list-item"
    >
      <div
        class="tui-competencyRatingScaleOverview__proficientStatus"
      >
        <checksuccess-stub
          size="300"
        />
      </div>
       
      <div
        class="tui-competencyRatingScaleOverview__scaleValue"
      >
        <h4
          class="tui-competencyRatingScaleOverview__scaleValue-name"
        >
          
          Not competent on Tuesdays
        
        </h4>
         
        <!---->
         
        <span
          class="sr-only"
        >
          
          $str('is_proficient_value', 'totara_competency')
        
        </span>
      </div>
    </li>
    <li
      class="tui-competencyRatingScaleOverview__list-item"
    >
      <div
        class="tui-competencyRatingScaleOverview__proficientStatus"
      >
        <!---->
      </div>
       
      <div
        class="tui-competencyRatingScaleOverview__scaleValue"
      >
        <h4
          class="tui-competencyRatingScaleOverview__scaleValue-name"
        >
          
          Below average
        
        </h4>
         
        <!---->
         
        <span
          class="sr-only"
        >
          
          $str('is_not_proficient_value', 'totara_competency')
        
        </span>
      </div>
    </li>
    <li
      class="tui-competencyRatingScaleOverview__list-item"
    >
      <div
        class="tui-competencyRatingScaleOverview__proficientStatus"
      >
        <!---->
      </div>
       
      <div
        class="tui-competencyRatingScaleOverview__scaleValue"
      >
        <h4
          class="tui-competencyRatingScaleOverview__scaleValue-name"
        >
          
          Not competent
        
        </h4>
         
        <!---->
         
        <span
          class="sr-only"
        >
          
          $str('is_not_proficient_value', 'totara_competency')
        
        </span>
      </div>
    </li>
  </ol>
</div>
`;
